### PR TITLE
Support column split in multi-target `hist`

### PR DESCRIPTION
**Pull request by rongou**
Also added support for categorical features in `approx`.

The main challenge is that the expand entry objects contain vector fields, which means the memory is not contiguous, so we cannot allgather the objects in one go. We have to allgather the primitive fields first, then separately allgather the vector fields.

An alternative may be to serialize the expand entry objects into a binary blob/json, do allgather, then deserialize them back. But that might have a bigger performance impact.

Future work may be to push `AllgatherV` into the communicator. I think we can implement it more efficiently in gRPC.

### CODE REVIEW

**Comment by rongou** [comment]
@trivialfis @hcho3 any idea on the test failures on macos-11? I can't reproduce it on my linux box.

**Comment by trivialfis** [comment]
Unfortunately, I don't have a macos device.

**Comment by hcho3** [comment]
Let me try to reproduce it on my Macbook.

**Comment by rongou** [comment]
I started a mac instance on AWS and was able to reproduce the test failures. I think I know where the failures are coming from, but still need to do some debugging to figure out a fix.

**Comment by rongou** [comment]
@trivialfis finally got the CI to pass. Please take a look.

File: src/common/quantile.cc [hunk 1]
```diff
@@ -371,7 +369,8 @@ auto AddCategories(std::set<float> const &categories, HistogramCuts *cuts) {
     InvalidCategory();
   }
   auto &cut_values = cuts->cut_values_.HostVector();
-  auto max_cat = *std::max_element(categories.cbegin(), categories.cend());
```
Comment (rongou) [inline src/common/quantile.cc:374]: @trivialfis I think this is a legit bug. If `categories` is empty, `std::max_element` returns `categories.cend`, which leads to undefined behavior when accessed.

Presumably with row-split `categories` would never be empty, but with column-split it's pretty common.

**Review by trivialfis — APPROVED** [review]
Looks good to me overall

File: src/common/quantile.cc (hunk 1 above)
Comment (rongou) [reply src/common/quantile.cc:374]: Done.

File: src/data/data.cc [hunk 2]
```diff
@@ -698,6 +698,9 @@ void MetaInfo::Extend(MetaInfo const& that, bool accumulate_rows, bool check_col
     this->feature_type_names = that.feature_type_names;
     auto &h_feature_types = feature_types.HostVector();
     LoadFeatureType(this->feature_type_names, &h_feature_types);
+  } else if (!that.feature_types.Empty()) {
```
Comment (trivialfis) [inline src/data/data.cc:701]: @rongou Could you please help comment on when the meta info has feature types but doesn't have type names?

Comment (rongou) [reply src/data/data.cc:701]: I think in some tests we set the feature types directly without setting the type names.

Comment (trivialfis) [reply src/data/data.cc:701]: Got it, let me try to run the tests.

**Pull request merged**